gh-89263: Add typing.get_overloads (#31716)
python/cpython: Doc/library/typing.rst

   See :pep:`484` for details and comparison with other typing semantics.

   .. versionchanged:: 3.11

>       Overloaded functions can now be retrieved at runtime :func:`get_overloads`.

Reviewer: "retrieved" sounds a bit unspecific. Maybe "introspected"? Also I'd add "using":
```suggestion
      Overloaded functions can now be introspected at runtime using :func:`get_overloads`.
```

.. function:: get_overloads(func)

   Return a sequence of :func:`@overload <overload>`-decorated definitions for *func*. *func* is
   the function object for the implementation of the overloaded function.
   For example, given the definition of ``process`` in the documentation for

>    :func:`@overload <overload>`, ``get_overloads(process)`` will return a sequence of three

Reviewer 2: Something about this feels off (to reference another doc example here). But I can't find a better way to illustrate your point.

>    Return a sequence of :func:`overload`-decorated definitions for *func*. *func* is
>    the function object for the implementation of the overloaded function.
>    For example, given the definition of ``process`` in the documentation for
>    :func:`overload`, ``get_overloads(process)`` will return a sequence of three
>    function objects for the three defined overloads.

Reviewer 3: Nit

```suggestion
   Return a sequence of :func:`@overload <overload>`-decorated definitions for *func*. *func* is
   the function object for the implementation of the overloaded function.
   For example, given the definition of ``process`` in the documentation for
   :func:`@overload <overload>`, ``get_overloads(process)`` will return a sequence of three
   function objects for the three defined overloads.
```

   See :pep:`484` for details and comparison with other typing semantics.

   .. versionchanged:: 3.11

>       Overloaded functions are now registered using :func:`functools.register_variant`.

Author: ```suggestion
      Overloaded functions can now be retrieved using :func:`functools.get_variants`.
```

   See :pep:`484` for details and comparison with other typing semantics.

   .. versionchanged:: 3.11
      Overloaded functions can now be introspected at runtime using
      :func:`get_overloads`.

.. function:: get_overloads(func)

>    Return a sequence of :func:`@overload <overload>`-decorated definitions for

Reviewer 3: Not a comment on this PR, but: how I wish there was a `:deco:` role that would add the `@` automatically, instead of having to do `:func:@overload <overload>` every time...

   *func*. *func* is the function object for the implementation of the

>    overloaded function. For example, given the definition of ``process`` in

Reviewer 3: This sentence sounds quite strange to me:

```
*func* is the function object for the implementation of the overloaded function.
```

But I'm not sure I have a better suggestion off the top of my head ://

Author: Yes, unfortunately the `overload` docs don't have a clear term for the implementation function either.

   the documentation for :func:`@overload <overload>`,
   ``get_overloads(process)`` will return a sequence of three function objects
   for the three defined overloads.

>    This function can be used for introspecting an overloaded function at

Reviewer 3: ```suggestion
   ``get_overloads`` can be used for introspecting an overloaded function at
```

The word "function" is used a lot in the previous paragraph

Author: It's all about functions. Applying this change locally.

Reviewer 3: > It's all about functions.

Well, of course :) that was just a rationale for the proposed change to *this* paragraph, not a criticism of the previous paragraph

Author: Sorry, I was joking. Anyway, I pushed the suggested changes, could you review again?

>    runtime.

Reviewer 3: ```suggestion
   runtime. If called on a function with no overloads, ``get_overloads``
   returns an empty sequence.
```

Author: Good point, but I'm adding it to the previous paragraph. The first paragraph is about what it does, this paragraph about what it's good for.

   See :pep:`484` for details and comparison with other typing semantics.

   .. versionchanged:: 3.11
      Overloaded functions can now be introspected at runtime using
      :func:`get_overloads`.

.. function:: get_overloads(func)

   Return a sequence of :func:`@overload <overload>`-decorated definitions for
   *func*. *func* is the function object for the implementation of the
   overloaded function. For example, given the definition of ``process`` in
   the documentation for :func:`@overload <overload>`,
   ``get_overloads(process)`` will return a sequence of three function objects
   for the three defined overloads. If called on a function with no overloads,

>    ``get_overloads`` returns an empty sequence.

Reviewer 3: ```suggestion
   return an empty sequence.
```

Tiny nit -- this paragraph starts off using the imperative mood

Author: But the previous sentence is in the indicative, so I think this is fine?

Reviewer 3: Well, it's not really a big deal either way :)
